### ISSUE TITLE
New Machine controller that powers OOT-providers has some regressions

### ISSUE THREAD

**Issue by zuzzas** [gardener/machine-controller-manager#522]
While attempting to integrate our new Yandex OOT driver for MCM. We aren't using Gardener, but these issues are unrelated.

- [x] Bootstrap token injection does not work. Fixed by https://github.com/gardener/machine-controller-manager/pull/521
- [ ] Machines are stuck in Pending even when the corresponding Node object is created and Running
- [x] The new Machine re-queuing mechanism is a step in the right direction. Unfortunately, it doesn't time-out Machines that are stuck on creation. Something might go wrong, and we'd like to recreate Machine any try again. Fixed by https://github.com/gardener/machine-controller-manager/pull/557. 

The new OOT controller is cobbled up from the parts on an old Machine controller. Phases are _not rigidly and clearly defined_ and it is hard to reason (with the current code) about what comes after or before what. Compare it to the cluster-api [approach](https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/machine_controller_phases.go#L49-L87).

I'd be happy to work on each of these issues, but I felt the need to aggregate my grievances into a single place.

**Comment by prashanth26** [comment]
Hi @zuzzas ,

Thank you for creating this master issue. With your tests, it really helps.

>  Machines are stuck in Pending even when the corresponding Node object is created and Running
I would be curious to know how I can reproduce this? 

> The new Machine re-queuing mechanism is a step in the right direction. Unfortunately, it doesn't time-out Machines that are stuck on creation. Something might go wrong, and we'd like to recreate Machine any try again.
If you mean by a backoff mechanism to re-create failing machines after some time (hopefully exponentially), this is something Hardik is actively working on. And we are planning to introduce a state called `CrashLookBackoff` that would re-try machine creations after failure (in the OOT we probably can change the way it works based on retry states as well).cc: @hardikdr  

> Phases are not rigidly and clearly defined and it has hard to reason (with the current code) about what comes after or before what. Compare it to the cluster-API approach.
I completely agree with this. I think it would make sense to create a documentation/state diagram to realize this? 

> I felt the need to aggregate my grievances into a single place.
Perfectly understandable. We can split and get the changes done.

**Comment by zuzzas** [comment]
>If you mean by a backoff mechanism to re-create failing machines after some time

`machine-creation-timeout` and its default value of 20 minutes does not work. Machines are stuck in the `Pending` state for eternity. Perhaps, they should not be deleted, but instead a combo of 

>I think it would make sense to create a documentation/state diagram to realize this?

Perhaps, not in a document form, but the code. Then we can generate a graph from the code, as some FSB libraries do. :)

Each phase should have multiple guard clauses (or checks, if you will). For example, for the Machine to transition into the Running phase it has to have a corresponding Node object and it has to be Running. A Machine should be able to transition into the Running phase only from Pending or Unknown phases. You can enter the Unknown phase from the Running state only. Only if Node became unhealthy for one reason or another. 

Basically, guard clauses and possible phases to enter from the current one.

**Comment by prashanth26** [comment]
> machine-creation-timeout and its default value of 20 minutes does not work. Machines are stuck in the Pending state for eternity (if they are not properly bootstrapped in the allotted time). Perhaps, they should not be deleted (they are moved to the Failed state and deleted in the old Machine controller), but instead a combo of DeleteMachine() and CreateMachine()?

I tried to reproduce the scenario of machines failing to join due to the bootstrap token generation issue. For the setup, i created machines whose bootstrap tokens have expired. The creation of the provider machine goes through but the machine joining times out. 

However, I see that the machines are being moved to the failed state (after creation timeout) and then are replaced by the machine-set controller as expected. At least machine replacement seems to be working as before to me. Did i miss something? Below are the logs.

```
I1001 05:58:27.872156       1 machine_util.go:640] Added finalizer to machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 05:58:27.872312       1 machine_controller.go:137] Machine status request has been received for "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 05:58:28.144022       1 machine_controller_util.go:366] machine name=shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl, uuid= not found
I1001 05:58:28.144097       1 machine.go:329] Creating a VM for machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl", please wait!
I1001 05:58:28.144114       1 machine_controller.go:65] Create machine request has been received for "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 05:58:34.480149       1 machine_controller.go:83] VM with Provider-ID: "gce:///sap-se-gcp-scp-k8s-dev/europe-west1-d/shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" created for Machine: "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 05:58:34.480220       1 machine.go:345] Created new VM for machine: "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" with ProviderID: gce:///sap-se-gcp-scp-k8s-dev/europe-west1-d/shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl
I1001 05:58:34.487751       1 machine.go:378] Machine labels/annotations UPDATE for "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 05:58:34.488082       1 machine_controller.go:137] Machine status request has been received for "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 05:58:34.802978       1 machine_controller_util.go:313] Found machine with name: "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 05:58:34.803300       1 machine_controller.go:154] Machine status: found VM "gce:///sap-se-gcp-scp-k8s-dev/europe-west1-d/shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" for Machine: "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 05:58:34.803427       1 machine.go:310] Found VM with required machine name. Adopting existing machine: "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" with ProviderID: gce:///sap-se-gcp-scp-k8s-dev/europe-west1-d/shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl
I1001 05:58:34.812733       1 machine.go:406] Machine/status UPDATE for "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" during creation
E1001 06:18:42.879706       1 machine_util.go:585] Machine shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl failed to join the cluster in 20m0s minutes.
I1001 06:18:42.889348       1 machine_util.go:612] Machine State has been updated for "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 06:18:42.913745       1 machine_util.go:722] Machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" status updated to terminating
I1001 06:18:42.914060       1 machine_controller.go:137] Machine status request has been received for "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 06:18:43.242785       1 machine_controller_util.go:313] Found machine with name: "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 06:18:43.242815       1 machine_controller.go:154] Machine status: found VM "gce:///sap-se-gcp-scp-k8s-dev/europe-west1-d/shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" for Machine: "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
E1001 06:18:43.259014       1 machine_util.go:808] Machine/status UPDATE failed for machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl". Retrying, error: Operation cannot be fulfilled on machines.machine.sapcloud.io "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl": the object has been modified; please apply your changes to the latest version and try again
W1001 06:18:43.259125       1 machine.go:497] Unable to decode deletion flow state for machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl". Re-initiate termination
I1001 06:18:43.266171       1 machine_util.go:722] Machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" status updated to terminating
I1001 06:18:43.266301       1 machine_controller.go:137] Machine status request has been received for "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 06:18:43.546397       1 machine_controller_util.go:313] Found machine with name: "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 06:18:43.546664       1 machine_controller.go:154] Machine status: found VM "gce:///sap-se-gcp-scp-k8s-dev/europe-west1-d/shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" for Machine: "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 06:18:43.565121       1 drain.go:169] Machine drain ended on 2020-10-01 06:18:43.565108886 +0000 UTC m=+5921.347194512 and took 6.851839ms for "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 06:18:43.565169       1 machine_util.go:932] Drain successful for machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl".
I1001 06:18:43.573634       1 machine_controller.go:102] Machine deletion request has been received for "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 06:18:43.862419       1 machine_controller_util.go:313] Found machine with name: "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 06:19:04.746565       1 machine_controller.go:114] VM "gce:///sap-se-gcp-scp-k8s-dev/europe-west1-d/shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" for Machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" was terminated succesfully
E1001 06:19:04.940134       1 machine_util.go:1100] Machine/status UPDATE failed for machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl". Retrying, error: Operation cannot be fulfilled on machines.machine.sapcloud.io "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl": the object has been modified; please apply your changes to the latest version and try again
I1001 06:19:05.137595       1 machine_util.go:663] Removed finalizer to machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl"
I1001 06:19:05.137627       1 machine.go:511] Machine "shoot--i341888--shoot-gcp-worker-1-z1-57fdcc6f48-l62gl" deleted successfully
```

**Comment by prashanth26** [comment]
>Perhaps, not in a document form, but the code. Then we can generate a graph from the code, as some FSM libraries do. :)

Each phase should have multiple guard clauses (or checks, if you will). For example, for the Machine to transition into the Running phase it has to have a corresponding Node object and it has to be Running. A Machine should be able to transition into the Running phase only from Pending or Unknown phases. You can enter the Unknown phase from the Running state only. Only if Node became unhealthy for one reason or another.

Basically, guard clauses and possible phases to enter from the current one.

Yes, sure I think that makes sense. To make the changing of phases more understandable through code itself. I like the idea.

**Comment by zuzzas** [comment]
@prashanth26 
Thanks for helping, I'll look further into it. I've tried to look through the OOT Machine controller's code, but the creation timeout is actually in the MachineSet controller? That's an oversight on my part, will look further into this.

**Comment by prashanth26** [comment]
Not really, the creation timeout is in the machine controller's code here - https://github.com/gardener/machine-controller-manager/blob/master/pkg/util/provider/machinecontroller/machine_util.go#L556-L598. And it sets the machine to a `Failed` state. And the machine-set controller tries to delete all `Failed` machines like it always used to.

On parallel thread - We are thinking to update the creation failure logic to allow retries of failed machine creations using probably a `CrashLoopBackOff` kind of phase where creation operations will be retried with an appropriate phase. This would help us tackling the constant retries (introduce backoff) on the failure scenarios. 

Yes, please do let me know if you are still facing issues.

**Comment by zuzzas** [comment]
Ah, I see. Perhaps, I am returning the wrong nodeName or providerID in the Yandex provider. I'll check, thanks!

Aye, that would be perfect. I've always disliked constant Machine object re-creations. And we are dealing with only one object that transitions from phase to phase, we can use [RateLimitingInterface](https://pkg.go.dev/k8s.io/client-go/util/workqueue#RateLimitingInterface) to implement [back-off](https://pkg.go.dev/k8s.io/client-go@v0.19.2/util/flowcontrol).

**Comment by prashanth26** [comment]
Hi @zuzzas ,

> Ah, I see. Perhaps, I am returning the wrong nodeName or providerID in the Yandex provider. I'll check, thanks!

Were you able to check why your machines weren't being deleted after a pending state? Just wanted to check with you on this.

**Comment by prashanth26** [comment]
Hi @zuzzas ,

Are any of the above mentioned issue still troubling you? If not i would like to close this issue.

**Comment by prashanth26** [comment]
2/3 of the above are fixed by mentioned PR in the issue description. 

> Machines are stuck in Pending even when the corresponding Node object is created and Running

Couldn't reproduce this one.
/close please reopen with details if you still face issues.
